src/dlls/mscoree/unixinterface.cpp

```diff
@@ -146,6 +138,61 @@ int coreclr_initialize(
     ConstWStringArrayHolder propertyValuesW;
     propertyValuesW.Set(StringArrayToUnicode(propertyCount, propertyValues), propertyCount);
 
+    // Process properties that specify CoreCLR startup flags and remove them from the property list
+    STARTUP_FLAGS startupFlags =
+        static_cast<STARTUP_FLAGS>(
+            STARTUP_FLAGS::STARTUP_LOADER_OPTIMIZATION_SINGLE_DOMAIN |
+            STARTUP_FLAGS::STARTUP_SINGLE_APPDOMAIN);
+    int newPropertyIndex = 0;
+    for (int oldPropertyIndex = 0; oldPropertyIndex < propertyCount; ++oldPropertyIndex)
+    {
+        if (wcscmp(propertyKeysW[oldPropertyIndex], W("CONCURRENT_GC")) == 0)
+        {
+            if (wcscmp(propertyValuesW[oldPropertyIndex], W("1")) == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | STARTUP_CONCURRENT_GC);
+            }
+        }
+        else if (wcscmp(propertyKeysW[oldPropertyIndex], W("SERVER_GC")) == 0)
+        {
+            if (wcscmp(propertyValuesW[oldPropertyIndex], W("1")) == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | STARTUP_SERVER_GC);
+            }
+        }
+        else if (wcscmp(propertyKeysW[oldPropertyIndex], W("HOARD_GC_VM")) == 0)
+        {
+            if (wcscmp(propertyValuesW[oldPropertyIndex], W("1")) == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | STARTUP_HOARD_GC_VM);
+            }
+        }
+        else if (wcscmp(propertyKeysW[oldPropertyIndex], W("TRIM_GC_COMMIT")) == 0)
+        {
+            if (wcscmp(propertyValuesW[oldPropertyIndex], W("1")) == 0)
+            {
+                startupFlags = static_cast<STARTUP_FLAGS>(startupFlags | STARTUP_TRIM_GC_COMMIT);
+            }
+        }
+        else
+        {
+            // This is not a CoreCLR startup flag, move the property to the new index if it's different from the old index
+            if (newPropertyIndex < oldPropertyIndex)
+            {
+                propertyKeysW[newPropertyIndex] = propertyKeysW[oldPropertyIndex];
```

Comment:
This will cause memory corruption on exit from this function - the ConstWStringArrayHolder will try to free the values at the end of the propertyKeysW twice.